subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java

```diff
@@ -159,29 +180,41 @@ public synchronized void close() {
             cacheUpdateExecutor.stop();
             cacheUpdateExecutor = null;
         }
+
+        withOwnership(() -> {
+            try {
+                crossProcessCacheAccess.close();
+                if (fileLockHeldByOwner != null) {
+                    fileLockHeldByOwner.run();
+                }
+
+                // If cleanup is required, but has not already been invoked (e.g. at the end of the build session)
+                // perform cleanup on close.
+                if (cleanupAction != null && cleanupAction.requiresCleanup() && !alreadyCleaned) {
+                    doCleanup();
+                }
+
+                if (cacheClosedCount != 1) {
+                    LOG.debug("Cache {} was closed {} times.", cacheDisplayName, cacheClosedCount);
+                }
+            } finally {
+                fileLockHeldByOwner = null;
+            }
+        });
+    }
+
+    void withOwnership(Runnable action) {
```

Comment:
I did this for the `open()` method and a few others.  This results in a slight reordering of a few things in `open()`, but I don't _think_ that should be an issue  There isn't any way to use this in the `useCache()` method because it behaves in a slightly different manner (i.e. taking ownership but not releasing it, then releasing it later) so I just left that one the way it is.  I don't _think_ I broke anything, but please take a good look at it and raise anything that might look suspicious.